query/filter/key_builder.go

```diff
@@ -162,15 +179,139 @@ func (s *StrictEqKeyComposer) Compose(selectors []*Selector, userDefinedKeys []*
 					return nil, errors.InvalidArgument("OR is not supported with composite primary keys")
 				}
 
-				key, err := s.keyEncodingFunc(sel.Matcher.GetValue().AsInterface())
+				primaryKeyParts := s.buildIndexPartsFunc(sel.Field.Name(), sel.Field.DataType, sel.Matcher.GetValue().AsInterface())
+
+				eqkey, err := s.keyEncodingFunc(primaryKeyParts...)
 				if err != nil {
 					return nil, err
 				}
 
+				key := newKeyType(EQUAL, schema.UnknownType, []keys.Key{eqkey})
 				allKeys = append(allKeys, key)
 			}
 		}
 	}
 
 	return allKeys, nil
 }
+
+const (
+	RANGE     string = "range"
+	FULLRANGE string = "full_range"
+	EQUAL     string = "equal"
+)
+
+// TODO: better name for this
+type KeyReadType struct {
+	QueryType string
+	DataType  schema.FieldType
+	Keys      []keys.Key
+}
+
+func newKeyType(queryType string, dataType schema.FieldType, keys []keys.Key) KeyReadType {
+	return KeyReadType{
+		queryType,
+		dataType,
+		keys,
+	}
+}
+
+// Range Key Composer will generate a range key set on the user defined keys
+type RangeKeyComposer struct {
+	// keyEncodingFunc returns encoded key from index parts
+	keyEncodingFunc     KeyEncodingFunc
+	buildIndexPartsFunc BuildIndexPartsFunc
+}
+
+func NewRangeKeyComposer(keyEncodingFunc KeyEncodingFunc, buildIndexParts BuildIndexPartsFunc) *RangeKeyComposer {
+	return &RangeKeyComposer{
+		keyEncodingFunc,
+		buildIndexParts,
+	}
+}
+
+func (s *RangeKeyComposer) Compose(selectors []*Selector, userDefinedKeys []*schema.QueryableField, parent LogicalOP) ([]KeyReadType, error) {
+	var err error
+	var foundKeys []KeyReadType
+	for _, k := range userDefinedKeys {
+		var begin, end keys.Key
+		rangeType := FULLRANGE
+		for _, sel := range selectors {
+			if k.FieldName == sel.Field.Name() && s.isRange(sel) {
+				indexParts := s.buildIndexPartsFunc(sel.Field.Name(), sel.Field.DataType, sel.Matcher.GetValue().AsInterface())
+				if s.isGreater(sel) {
```

Comment:
branches looks similar, maybe generalize, or at least move to their own functions.